OS X improvements

techee opened the pull request:
I have finally had some time to play with geany on OS X and found a few problems these patches try to address. For my installation I used the quartz backend as described here: http://wiki.geany.org/howtos/osx/running

The biggest issue is that paste doesn't work - the issue is in Scintilla (well, rather in gtk-quartz but it's easier to workaround in Scintilla). I've already submitted the patch here: https://sourceforge.net/p/scintilla/bugs/1677/

The second biggest problem is the infinite loop in find in files which should be fixed now.

The rest are minor problem fixes and fixes of warnings comping from clang.

One more patch to mention is the ui_get_mime_icon() improvements - it fixes the case (which is a bit surprising) when GIcon is returned but the icon actually doesn't exist (haven't seen this happen on Linux, but it happens on OS X and I believe on Windows too). It also replaces the "stock ID" (which for some reason doesn't work) with "icon name" which works (e.g. for folder icons). I haven't found since which GTK version icon names are supported so I hope it works on older GTK too.

After the patches, Geany is quite usable (and nice-looking) on OS X - the only issue I'm aware of is the non-functional VTE terminal (terminal tab shows but it's empty all the time). But I suspect VTE here - I compiled libvte and tried the supplied demo vte application and it doesn't work either.

eht16 commented:
I just tested the icon changes in https://github.com/techee/geany/commit/7880c0fc123da98ffa0881f84c93714c60526ac6 on Windows and they fix our known problems there as well (as described in http://lists.geany.org/devel/2014-October/009182.html).

techee commented:
I've just re-pushed the patches addressing various Colomban's comments. At the end I left the Scintilla-buffer-disabling patch without the dynamic checks - these checks aren't available on GTK2 so there would have to be additional ifdefs and it didn't seem worth it - the flag settings isn't so critical and Scintilla will work fine even when it's disabled and using X.

The copy/paste patch has already been applied to Scintilla so the next time Scintilla is updated, Geany will get the fix.

Finally there's one extra patch dealing with keybindings. On OS X the Command key is normally used instead of Ctrl so the keybindings have to be updated accordingly. There were some complications on the way - see the patch for more details.

b4n commented:
> The copy/paste patch has already been applied to Scintilla so the next time Scintilla is updated, Geany will get the fix.

You (or I) can import the Scintilla patch from Scintilla HG in here if you want.

techee commented:
> The copy/paste patch has already been applied to Scintilla so the next
> time Scintilla is updated, Geany will get the fix.
>
> You (or I) can import the Scintilla patch from Scintilla HG in here if you
> want.
>

Good. I'll include it the next time I push the osx patches.

techee commented:
I have re-pused the patches again. As we discussed, I've changed the keybindings patch to use the keybindings_get_modifiers() function and a single key mask macro. I have split this patch into 2 parts - the plugin API part is separate in case it conflicts with Matthew's work - it can be applied later if needed.

There is one extra patch which sets better default fonts on OSX.

techee commented:
OK, two more fixes but I think that's about it - everything works pretty well now (except VTE, printing and possibly tons of things I haven't tested).

techee commented:
@codebrainz Have you tried to create the bundle when playing with Geany on OS X? If so, do you have some config files or scripts for this? It would be nice to create a proper Geany bundle but I guess it requires quite a lot of configuration of what should be included into the bundle so if you have something, I'd be definitely interested to have a look.

codebrainz commented:
@techee yeah I did have some scripts to make the bundle, I think it was based on the stuff in gtk-osx-application, which I also used to make a proper main menu in the top panel and stuff. I'll try and find it on my macbook.

codebrainz commented:
@techee seem I already had it pushed, it's been a while, but most of the bundle making stuff is in `osx/` directory IIRC. https://github.com/codebrainz/geany/tree/wip/gtk-osx

techee commented:
Cool, thanks. I'll definitely have a look (once I summon some motivation to do this annoying config file work).

codebrainz commented:
@techee for that, particularly [this](https://github.com/codebrainz/geany/blob/dd5c485007c36abef71013227246414c5f142b30/osx/geany.conf) for config and [this](https://github.com/codebrainz/geany/blob/wip/gtk-osx/osx/keybindings.conf) for keybindings, in that case.

b4n commented:
Hum, I see I introduced a merge conflict in 91d5079c8d56c69d78cad6d4cb9600b89a86d085, sorry :/

techee commented:
No worries, it's easy to fix.

techee commented:
I have repushed the patches again:

It's rebased on top of current master so there should be no conflict now.

I've moved the docstrings from the "fix keybindings" patch to "make keybindings_get_modifiers() part of API" and replaced "valid" with "significant" (it's used in the _set_default_mask() documentation and I think it's better than "interesting").

I used single quotes for the "cd" command in the terminal execution script and g_strescape() for the directory which should be sufficient (we actually need to escape just single quotes in this case).

I have left the "better fonts" patch there - feel free to drop it if you don't like it, I just think it's better to provide some useable defaults.

techee commented:
Since Scintilla 3.5.3 is in master now, I have removed the Scintilla "paste patch" and rebased on top of master.

b4n commented:
OK, looks great to me.

@codebrainz had you a chance to give this a try on OSX, and if not, would you mind giving it a quick shot to see if nothing obvious has been overlooked? -- or is machine/version dependent.

b4n commented:
> @techee for that case, particularly [this](https://github.com/codebrainz/geany/blob/dd5c485007c36abef71013227246414c5f142b30/osx/geany.conf) for config

I must say I love this approach (especially for fonts) :)  Yet, I also understand one would like something clean by default -- but what about we install a default configuration?

techee commented:
Note that Matthew's config file is used just when you create a bundle (I haven't tried that yet) but not when you install geany from sources or some port distribution like macports or homebrew. We could use different default config files for each of the platforms we support but I'm not sure all the automake/autoconf/waf work is worth it (detecting the correct platform, checking whether there's a config file already and copying the right config file).

The fonts I used correspond to the standard osx monospace and user interface fonts (the ones used by pango are wrong - these are just some fallback platform-independent fonts). In the future we could introduce something similar to win32.c/h for os x (including e.g. native file open dialogs) and we could use osx API to query for default fonts there. So I think it's better to do this in the code.

@codebrainz If you want to test it quickly, just follow http://wiki.geany.org/howtos/osx/running and then uninstall geany from macports by "sudo port install geany". Then you can install geany from sources having all the dependencies installed.

b4n commented:
> We could use different default config files for each of the platforms we support but I'm not sure all the automake/autoconf/waf work is worth it (detecting the correct platform, checking whether there's a config file already and copying the right config file).

You may have a point here indeed.

> The fonts I used correspond to the standard osx monospace and user interface fonts (the ones used by pango are wrong - these are just some fallback platform-independent fonts).

OK

> In the future we could introduce something similar to win32.c/h for os x (including e.g. native file open dialogs)

I'd rather not if possible to avoid :)

> and we could use osx API to query for default fonts there.

Well, hopefully Pango should be fixed rather than each application, shouldn't it?

techee commented:
>  In the future we could introduce something similar to win32.c/h for os x
> (including e.g. native file open dialogs)
>
> I'd rather not if possible to avoid :)
>
I understand it can be problematic to maintain when there's just one person
who has the right hardware and knows the API (and who can lose interest or
not have time to keep maintaining it) but I was thinking about having
functions which would always have a gtk fallback mode, e.g.:

void show_open_file_dialog()
{
#ifdef USE_GTK
    //open using gtk
#else
    //open using os x api
#endif
}

so there would always be the option to drop the platform-specific code.

But don't worry - I probably won't have time for this in the near future
:-).


>  and we could use osx API to query for default fonts there.
>
> Well, hopefully Pango should be fixed rather than each application,
> shouldn't it?
>
Yes, I might have a look at it but even if it gets fixed now, geany can't
rely on this in the near future because there would still be many versions
in the wild without the fix.

b4n commented:
> but I was thinking about having functions which would always have a gtk fallback mode, e.g.: […]
> so there would always be the option to drop the platform-specific code.

We have that for Windows, but then we are reluctant to remove it because it'd annoy people and the like :)

> Yes, I might have a look at it but even if it gets fixed now, geany can't rely on this in the near future because there would still be many versions in the wild without the fix.

Sure, it wouldn't be a "right now" solution and would need time to settle.  I'm just trying to avoid platform-specific code in Geany, so if Pango gets fixed before anyone starts adding custom code it's all good :)

techee commented:
> but I was thinking about having functions which would always have a gtk
> fallback mode, e.g.: […]
> so there would always be the option to drop the platform-specific code.
>
> We have that for Windows, but then we are reluctant to remove it because
> it'd annoy people and the like :)
>
Yes, I know that very well. Stupid people :-).

codebrainz commented:
On 15-01-21 11:19 AM, Jiří Techet wrote:
> Note that Matthew's config file is used just when you create a bundle (I haven't tried that yet) but not when you install geany from sources or some port distribution like macports or homebrew. [...]

I'm pretty sure these channels are only really useful for projects that 
don't provide a proper OSX release. That was kind of the goal with the 
work on the bundle was to provide a normal way to install Geany on OSX. 
Sadly I find OSX to be a pretty awful operating system overall (compared 
to Linux), so I rarely use it for more than watching youtube videos, so 
I doubt I'll be pushing forward to get us releasing an OSX bundle any 
time soon :)

techee commented:
I've just added one more patch which gets rid of the __sentinel__ warning we talked about before. It probably is a clang bug so instead of removing the G_GNUC_NULL_TERMINATED macro, just disable the corresponding warning for this header when using clang.

techee commented:
Yet another minor fix - "echo -n" does't do the right thing. The -n switch is implementation-dependent so avoid it:

http://pubs.opengroup.org/onlinepubs/009696799/utilities/echo.html

techee commented:
Just nitpicking - CR line ending isn't used since OS X and calling it just "Mac" line ending is a bit misleading. Clarify this a bit.

codebrainz commented:
`"Classic _Mac"`

techee commented:
Yes, I think you are right, I used the Sublime Text wording but the majority of OS X editors uses "Classic Mac":

Sublime Text: Mac OS 9
Xcode: Classic Mac OS
Textmate: Mac Classic
IntelliJ: Classic Mac
BBEdit: Classic Mac
Eclipse: haven't installed ;-)

techee commented:
Updated the patches.

About the clang flags: I would personally prefer to have Geany compiled by clang without warnings by default because one has to set the flag from the command line including the flags Geany normally uses - I can do that, it's just annoying.

Anyway, I have removed it from the OS X patches and hopefully if Matthew finishes his symbols work it won't be needed any more.

techee commented:
I added one more patch to check for VTE with dylib extension (this one is used on OS X instead of .so). I didn't include this one originally because VTE doesn't work with GTK 2 but I realized it works with GTK 3 so the patch makes sense.

Yeah and by the way, the latest VTE (libvte2_91.so) doesn't load because vte_terminal_fork_command() was removed and there's just vte_terminal_fork_command_full() now.

b4n commented:
> About the clang flags: I would personally prefer to have Geany compiled by clang without warnings by default because one has to set the flag from the command line including the flags Geany normally uses - I can do that, it's just annoying.

If disabling it is useful, I think I'd prefer to have the build system detect clang and disable the flag here -- it's not like this flag is dead important, and GCC is IMHO the canonical reference for those anyway.

> Anyway, I have removed it from the OS X patches and hopefully if Matthew finishes his symbols work it won't be needed any more.

Indeed it would nullify the issue :)

b4n commented on data/geany.glade:
```diff
@@ -7726,7 +7726,7 @@
                                     <property name="visible">True</property>
                                     <property name="can_focus">False</property>
-                                    <property name="label" translatable="yes">Convert and Set to CR (_Mac)</property>
+                                    <property name="label" translatable="yes">Convert and Set to CR (_Mac OS 9)</property>
```
Didn't you mean to label these "Classic Mac" or something? https://github.com/geany/geany/pull/396#issuecomment-73346068 https://github.com/geany/geany/pull/396#issuecomment-73358735

techee replied:
Crap, what have I pushed? I really need to be reviewed, thanks :-).

b4n commented:
> Yeah and by the way, the latest VTE (libvte2_91.so) doesn't load because vte_terminal_fork_command() was removed and there's just vte_terminal_fork_command_full() now.

Yeah 2.91 was an API break and we'd need porting, but I tried and it required some subtle changes.  #336

techee commented:
Ouch, the VTE API change is much bigger than I thought.

b4n commented:
As someone else seems to have tested this (see ML), I merged it -- we can always fix possible issues later.
